tools/topology/sof/pipe-passthrough-playback-variable.m4

```diff
@@ -0,0 +1,46 @@
+# Low Latency Passthrough Pipeline and PCM
+#
+# Pipeline Endpoints for connection are :-
```

Comment:
@juimonen PR looks good to me, I have a few questions in order to clarify some things. 

Feel free to answer them and also annotate the commit message if you find it useful.

1) What is passthrough pipeline in this context? Since m4 files are hard to read. It looks like this passthrough pipeline is just a pipeline with just host and DAI.
2) Who sets the number of channels and the format? Is this set at runtime by parameters taken from aplay command line for example?
3) Can the same be done for sample rate? E.g  I want to set a range of rates at topology load and then at runtime to just select one of them based on the actual `wav` file.

Reply:
@dbaluta yes all these changes are related to using multiple dai configuration. So exactly we could change for example SSP config based on sample rate (or some other rule).

1) yes correct. Purpose of this new pipeline is that I just want to change the channel count with pcm from 1 -> PIPELINE_CHANNELS. We seem to have all pipeline channel counts hardcoded to 2. So if you use this pipeline, aplay in user space can play mono and stereo (if you set the pipeline channels to 2 in top level -> so it will be range from 1-2). Not sure we had ever before played mono pipelines, but it seem to work fine. So I agree it is confusing -> basically I set min channel count in this pipe-variable file for the pcm and max channel count is coming from top level topology pipeline channel count. Would be nice if min and max channels would come from top level, but I really don't want to start changing this, because it will change all our pipelines -> error prone at this point.

2) yes. I'm trying here to define "ranges" for both channel count and format. For sample rate we are already doing that (as we have src in some pipelines). So this means that you need to make sure your pipeline supports this. The pipelines get the pcm hw params when streaming starts and they are propagated through the pipeline to the dai. So let's say your DAI supports 32 bit data: if the pcm format range supports playing 16 bit and you try to play 16 bit data -> the format is propagated to dsp and only when it hits the dai -> you get error that this is not supported. If we limit this format in topology to 32bit and you try to play 16bit -> everything fails already before hitting dsp: there are checks already in the alsa driver level.

3) yes we are now trying to use this in a case to change the SSP config based on sample rate:
https://github.com/thesofproject/linux/pull/2788

So sorry for the long rant: the new passthrough pipeline is mainly done to let us use variable _ranges_ so that alsa driver doesn't stop us before reaching dsp. The example topology shows how to define 2 SSP configs for a single topology. And you need some kernel changes PR2788 to somebody actually _change_ the dai config based on some rule (this case sample rate). And all this is because we have use case, where we have a chip/codec behind SSP, of which the config is changing based on use case (sample rate in this case).

Reply:
@juimonen thanks! This is a thing that @ViorelSuman is looking at. We want to be able to test our hardware DAIs / DACs with all their advertised parameters (e.g rates, channels, width) but ideally without changing the topology and also without a rate converter.

Reply:
@ViorelSuman any comments here ? Does this work for you ?

Reply:
@lgirdwood all good.